src/fhirtypes/StructureDefinition.ts

```diff
@@ -341,37 +337,71 @@ export class StructureDefinition {
-    // See: https://blog.fire.ly/2019/09/13/type-slicing-in-fhir-r4/
-    j.differential.element = j.differential.element.filter(
-      (e: { id: string; path: string; slicing: ElementDefinitionSlicing }) => {
-        return (
-          // not a choice, or
-          !e.id.endsWith('[x]') ||
-          // is a choice but not one whose only diff is the standard type slicing
-          !(
-            Object.keys(e).length === 3 &&
-            e.id &&
-            e.path &&
-            e.slicing?.discriminator?.length === 1 &&
-            e.slicing.discriminator[0].type === 'type' &&
-            e.slicing.discriminator[0].path === '$this' &&
-            e.slicing.rules === 'open' &&
-            (e.slicing.ordered == null || e.slicing.ordered === false)
-          )
-        );
+    // Populate the differential
+    j.differential = { element: [] };
+    this.elements.forEach(e => {
+      let hasDiff = cachedHasDiff(e);
+      let choiceSliceRoot = false;
+      if (hasDiff) {
+        const diff = e.calculateDiff().toJSON();
+        // choice[x] elements that differ only by the standard type slicing definition don't need to go in the
+        // differential, as they are inferred. See: https://blog.fire.ly/2019/09/13/type-slicing-in-fhir-r4/
+        if (
+          diff.id.endsWith('[x]') &&
+          Object.keys(diff).length === 3 &&
+          diff.id &&
+          diff.path &&
+          diff.slicing?.discriminator?.length === 1 &&
+          diff.slicing.discriminator[0].type === 'type' &&
+          diff.slicing.discriminator[0].path === '$this' &&
+          diff.slicing.rules === 'open' &&
+          (diff.slicing.ordered == null || diff.slicing.ordered === false)
+        ) {
+          // treat it like a no-diff element
+          hasDiff = false;
+          choiceSliceRoot = true;
+        } else {
+          j.differential.element.push(diff);
+        }
+      }
+
+      // Since the publisher can be buggy w/ sparse differentials, include a simple representation of
+      // the element if it has any children that will be in the differential. Since the children()
+      // function doesn't include slices, also check the slices since we want an element representing
+      // the slice root if there are slices (BTW - slices will always be hasDiff===true since we always put
+      // the sliceName in the diff). There is one exception: if this element is a choice[x] only differing
+      // by the standard choice slicing *and* it doesn't have any direct children w/ diffs, don't include
+      // it even if it has slices because it makes the diff a bit more confusing and it is supposed to
+      // be inferred anyway (see blog reference above).
+      if (
+        !hasDiff &&
+        (e.children().some(c => cachedHasDiff(c)) ||
+          (!choiceSliceRoot && e.getSlices().some(c => cachedHasDiff(c))))
+      ) {
+        j.differential.element.push({
+          id: e.id,
+          path: e.path
+        });
```

Comment:
I think it's needed to handle the case where there is a choice, which is then sliced, *but* there is also a constraint on a child of the choice itself.  E.g.,

```
* value[x].extension ^short = "an extension on any of the choices"
* valueQuantity units = UCUM#mg
```

In your version of the code, I think there would not be an element for `value[x]` even though there is an element for `value[x].extension`.  I can add a unit test to confirm.

Reply:
I've added a test and confirmed that your code above causes the test to fail (for the reason outlined in my comment above).  That said, I refactored the code to avoid the confusing extra `if`.  It causes a small amount of duplicate code, but is probably worth the cost in readability.